### PR TITLE
Clarify maxExportBatchSize behavior for Periodic exporting MetricReader

### PR DESCRIPTION
**Pull request by dashpole**
Part of https://github.com/open-telemetry/opentelemetry-specification/issues/5184

## Changes

Clarifies the behavior of `maxExportBatchSize` on `Periodic exporting MetricReader` based on feedback and findings from prototype implementations:
- Clarified that `exportTimeoutMillis` applies to each individual `Export(batch)` invocation when `maxExportBatchSize` splits collected metric data into multiple batches.
- Clarified that batches produced from a single `Collect()` call must be provided to `Export` serially and in-order before metric points from subsequent collections are provided.
- Clarified that batches produced by splitting should not contain empty metric structures (metrics containing zero metric data points).
- Clarified that if an export is still in progress when the next scheduled interval occurs, the reader must delay subsequent collection/export or skip that interval to prevent concurrent `Export` calls.
- Clarified that `ForceFlush` exports batches serially and returns an error status if any `Export(batch)` call or the exporter's `ForceFlush()` fails/times out.
* [x] Related issues: #5184
* [ ] Related [OTEP(s)](https://github.com/open-telemetry/oteps) #
* [x] Links to the prototypes (when adding or changing features):
  * Go: https://github.com/open-telemetry/opentelemetry-go/pull/8071
  * Java: https://github.com/open-telemetry/opentelemetry-java/pull/8296
  * JS: https://github.com/open-telemetry/opentelemetry-js/pull/6655
* [x] [`CHANGELOG.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md) file updated for non-trivial changes
* [ ] [Spec compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix/template.yaml) updated if necessary
* [ ] [Declarative config data model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/data-model.md#overview) is updated if SDK config surface is changed

### CODE REVIEW

File: specification/metrics/sdk.md [hunk 1]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
...
-ensure all metric data points from a single `Collect()` are provided to
-`Export` before metric data points from a subsequent `Collect()` so that metric
-points are sent in-order. The reader MUST NOT combine metrics from different
-`Collect()` calls into the same batch provided to `Export`.
+ensure all batches produced from a single `Collect()` are provided to `Export`
+serially and in-order before metric data points from a subsequent `Collect()`
+are provided. The reader MUST NOT combine metrics from different `Collect()`
+calls into the same batch provided to `Export`. Batches produced by splitting
+SHOULD NOT contain empty metric structures (metrics containing zero metric data
+points).
```
[the remark this answers is not in the document]
Comment (dashpole) [reply specification/metrics/sdk.md]: this is SHOULD NOT for backwards-compatibility.

File: specification/metrics/sdk.md [hunk 2]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
...
 
 The reader MUST synchronize calls to `MetricExporter`'s `Export`
-to make sure that they are not invoked concurrently.
+to make sure that they are not invoked concurrently. If an export is still in
+progress when the next scheduled interval occurs, the reader MUST either delay
+the subsequent collection and export until the in-progress export finishes, or
+skip the scheduled collection for that interval.
```
[the remark this answers is not in the document]
Comment (dashpole) [reply specification/metrics/sdk.md:1574]: I left this open because I suspect languages will want flexibility in how they handle this. Language implementations already have behavior, so this shouldn't be too prescriptive

File: specification/metrics/sdk.md [hunk 3]
```diff
@@ -1588,18 +1596,19 @@ This method provides a way for the periodic exporting MetricReader
...
 `ForceFlush` SHOULD provide a way to let the caller know whether it succeeded,
-failed or timed out. `ForceFlush` SHOULD return some **ERROR** status if there
-is an error condition; and if there is no error condition, it should return some
-**NO ERROR** status, language implementations MAY decide how to model **ERROR**
-and **NO ERROR**.
+failed or timed out. If any [`Export(batch)`](#exportbatch) call fails or times
+out, or if the configured exporter's [`ForceFlush()`](#forceflush-2) fails or
+times out, `ForceFlush` SHOULD return some **ERROR** status. If all calls
+succeed, it should return some **NO ERROR** status. Language implementations
+MAY decide how to model **ERROR** and **NO ERROR**.
```
[the remark this answers is not in the document]
Comment (dashpole) [reply specification/metrics/sdk.md]: addressed in 51dcdd00

File: CHANGELOG.md [hunk 4]
```diff
@@ -13,6 +13,10 @@ release.
 
 ### Metrics
 
+- Clarify `maxExportBatchSize` behavior, timeouts, and error handling for
+  Periodic exporting MetricReader.
+  ([#5184](https://github.com/open-telemetry/opentelemetry-specification/issues/5184))
```
Comment (reyang) [inline CHANGELOG.md]:
```suggestion
  ([#5265](https://github.com/open-telemetry/opentelemetry-specification/pull/5265))
```

**Review by reyang — APPROVED** [review]

**Review by jack-berg — APPROVED** [review]

File: specification/metrics/sdk.md [hunk 5]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
...
 points that belong to the same metric into different batches. The reader MUST
-ensure all metric data points from a single `Collect()` are provided to
-`Export` before metric data points from a subsequent `Collect()` so that metric
-points are sent in-order. The reader MUST NOT combine metrics from different
-`Collect()` calls into the same batch provided to `Export`.
+ensure all batches produced from a single `Collect()` are provided to `Export`
+serially and in-order before metric data points from a subsequent `Collect()`
+are provided. The reader MUST NOT combine metrics from different `Collect()`
+calls into the same batch provided to `Export`. Batches produced by splitting
+SHOULD NOT contain empty metric structures (metrics containing zero metric data
```
Comment (cijothomas) [inline specification/metrics/sdk.md]: nit qn: Do we need to explicitly say batch should not contain metric with zero metric points? What situation would cause an implementation to produce it?

**Review by cijothomas — APPROVED** [review]
LGTM, left a non-blocking nit suggestion.

File: specification/metrics/sdk.md (hunk 5 above)
Comment (dashpole) [reply specification/metrics/sdk.md]: It was a bug that was uncovered by reviewers in two (java and node js) of the three prototype implementations.  I agree it isn't strictly necessary, and is just an edge case implementations should already know to test for.  LMK if you would prefer not to have it.

Comment (reyang) [reply specification/metrics/sdk.md]: > It was a bug that was uncovered by reviewers in two (java and node js) of the three prototype implementations

Curious, why is this considered a bug? I can think of couple scenarios where one would like to distinguish between "I have no metric points and I don't know if that's because of a telemetry delay or my HTTP server has received no incoming request" and "I have no metric points and I know that it is because my HTTP server has received no incoming request, it must not be my telemetry system delay". Do you see those requirements?

Comment (dashpole) [reply specification/metrics/sdk.md]: It isn't about avoiding empty batches entirely. E.g. if the splitter is provided a batch that exactly matches the batch size, it shouldn't split into a full batch, and a zero-sized batch.  It should only include the full batch.  Or another example would be if all the datapoints from a metric are split out of a batch into a separate one, it shouldn't send the metric without any datapoints in the second batch.

Comment (reyang) [reply specification/metrics/sdk.md]: Got it, thanks!

**Review by pellared — APPROVED** [review]

File: specification/metrics/sdk.md [hunk 6]
```diff
@@ -1588,18 +1596,19 @@ This method provides a way for the periodic exporting MetricReader
...
 `ForceFlush` SHOULD collect metrics, split into batches if necessary, call
-[`Export(batch)`](#exportbatch) on each batch and
+[`Export(batch)`](#exportbatch) on each batch serially, and call
 [`ForceFlush()`](#forceflush-2) on the configured
 [Push Metric Exporter](#push-metric-exporter). `ForceFlush` MAY skip
 [`Export(batch)`](#exportbatch) calls if the timeout is already expired, but
 SHOULD still call [`ForceFlush()`](#forceflush-2) on the configured
 [Push Metric Exporter](#push-metric-exporter) even if the timeout has passed.
 
 `ForceFlush` SHOULD provide a way to let the caller know whether it succeeded,
```
Comment (dashpole) [inline specification/metrics/sdk.md]: New `ForceFlush` error enumeration contradicts the adjacent timeout paragraph and drops collection failures. Keep the enumeration illustrative rather than exhaustive, and keep the.

File: specification/metrics/sdk.md [hunk 7]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
...
+ensure all batches produced from a single `Collect()` are provided to `Export`
+serially and in-order before metric data points from a subsequent `Collect()`
+are provided. The reader MUST NOT combine metrics from different `Collect()`
+calls into the same batch provided to `Export`. Batches produced by splitting
+SHOULD NOT contain empty metric structures (metrics containing zero metric data
+points).
 
 The reader MUST synchronize calls to `MetricExporter`'s `Export`
-to make sure that they are not invoked concurrently.
+to make sure that they are not invoked concurrently. If an export is still in
```
Comment (dashpole) [inline specification/metrics/sdk.md]: New unconditional scheduling `MUST` in Stable text narrows previously-permitted implementations. Either (a) move the sentence into the `Status: [Development]`.

File: specification/metrics/sdk.md [hunk 8]
```diff
@@ -1548,7 +1548,10 @@ Configurable parameters:
 * `exportIntervalMillis` - the time interval in milliseconds between two
   consecutive collections. The default value is 60000 (milliseconds).
 * `exportTimeoutMillis` - how long the export can run before it is cancelled.
```
Comment (dashpole) [inline specification/metrics/sdk.md]: Per-batch export timeout makes total export duration for one collection unbounded, and the bullet is missing its Development marker. Add the Development marker, and bound the aggregate.

File: specification/metrics/sdk.md [hunk 9]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
 batches. The initial batch of metric data MUST be split into as many "full"
```
Comment (dashpole) [inline specification/metrics/sdk.md]: Error handling on the periodic (non-ForceFlush) export path is still unspecified. Add one sentence to the Development paragraph, e.g.

File: specification/metrics/sdk.md [hunk 10]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
...
 batches of size `maxExportBatchSize` as possible -- even if this splits up data
 points that belong to the same metric into different batches. The reader MUST
-ensure all metric data points from a single `Collect()` are provided to
-`Export` before metric data points from a subsequent `Collect()` so that metric
-points are sent in-order. The reader MUST NOT combine metrics from different
-`Collect()` calls into the same batch provided to `Export`.
+ensure all batches produced from a single `Collect()` are provided to `Export`
+serially and in-order before metric data points from a subsequent `Collect()`
+are provided. The reader MUST NOT combine metrics from different `Collect()`
+calls into the same batch provided to `Export`. Batches produced by splitting
```
Comment (dashpole) [inline specification/metrics/sdk.md]: "SHOULD NOT contain empty metric structures" is ambiguous, uses undefined terminology, and contradicts the linked Go prototype's pinned behavior. Say what happens, at all three levels, and answer the whole-batch case.

File: specification/metrics/sdk.md [hunk 11]
```diff
@@ -1548,7 +1548,10 @@ Configurable parameters:
 * `exportIntervalMillis` - the time interval in milliseconds between two
   consecutive collections. The default value is 60000 (milliseconds).
 * `exportTimeoutMillis` - how long the export can run before it is cancelled.
-  The default value is 30000 (milliseconds).
+  The default value is 30000 (milliseconds). When `maxExportBatchSize` is
+  configured and results in splitting the collected metric data into multiple
+  batches, `exportTimeoutMillis` applies to each individual `Export(batch)`
+  invocation.
 * **Status**: [Development](../document-status.md) - `maxExportBatchSize` - the
```
Comment (dashpole) [inline specification/metrics/sdk.md]: No default and no invalid-value behavior documented for `maxExportBatchSize`, though the declarative config schema already defines both. "The default is to not limit the batch size (no splitting).

File: specification/metrics/sdk.md (hunk 9 above)
Comment (dashpole) [inline specification/metrics/sdk.md]: Splitting does not say that `Resource` and `Scope` are repeated in every produced batch. "Each batch produced by splitting MUST carry the same `Resource` as the.

Comment (dashpole) [inline specification/metrics/sdk.md]: Partial-batch failure and its temporality consequences are unaddressed. A non-normative note: "Note: when splitting is in effect, an export.

File: specification/metrics/sdk.md [hunk 12]
```diff
@@ -1548,7 +1548,10 @@ Configurable parameters:
 * `exportIntervalMillis` - the time interval in milliseconds between two
```
Comment (dashpole) [inline specification/metrics/sdk.md]: `Shutdown`'s interaction with `maxExportBatchSize` is left undefined while `ForceFlush`'s is now specified. One sentence in the Development paragraph noting that the same splitting,.

File: specification/metrics/sdk.md [hunk 13]
```diff
@@ -1559,13 +1562,18 @@ configured, the reader MUST ensure no batch provided to `Export` exceeds the
 batches. The initial batch of metric data MUST be split into as many "full"
 batches of size `maxExportBatchSize` as possible -- even if this splits up data
 points that belong to the same metric into different batches. The reader MUST
```
Comment (dashpole) [inline specification/metrics/sdk.md]: "in-order" within a single `Collect()` is a `MUST` without an interoperability justification. Low stakes — implementations get in-order for free.

File: specification/metrics/sdk.md (hunk 5 above)
Comment (dashpole) [reply specification/metrics/sdk.md]: decided to remove the sentence in 1ea218d4
